packages/compiler/block.ts

```diff
@@ -1,51 +1,288 @@
+    t.isVariableDeclarator(callExpressionPath.parent) &&
+    t.isIdentifier(callExpressionPath.parent.id)
+      ? callExpressionPath.parent.id
+      : null;
+
+  if (
+    !t.isIdentifier(rawComponent) &&
+    !t.isFunctionDeclaration(rawComponent) &&
+    !t.isFunctionExpression(rawComponent) &&
+    !t.isArrowFunctionExpression(rawComponent)
+  ) {
+    throw deopt(
+      'Unsupported argument for block. Make sure blocks consume a component ' +
+        'function or component reference',
+      info.filename,
+      callExpressionPath
+    );
+  }
+
+  if (t.isIdentifier(rawComponent) && info.blocks.has(rawComponent.name)) {
+    callExpressionPath.replaceWith(rawComponent);
+    return;
+  }
+
+  /**
+   * Replaces `export default block(Component)` with
+   * const default$ = block(Component);
+   * export default default$;
+   */
+  const exportDefaultPath = callExpressionPath.parentPath;
+  if (exportDefaultPath.isExportDefaultDeclaration()) {
+    let name: string;
+    if (t.isIdentifier(rawComponent)) {
+      name = rawComponent.name;
+    } else if (
+      t.isFunctionExpression(rawComponent) &&
+      t.isIdentifier(rawComponent.id)
+    ) {
+      name = rawComponent.id.name;
+    } else {
+      name = 'default$';
+    }
+
+    const exportId = getUniqueId(info.programPath, name);
+    const paths = exportDefaultPath.insertBefore(
+      t.variableDeclaration('let', [
+        t.variableDeclarator(exportId, callExpression),
+      ])
+    );
+    exportDefaultPath.replaceWith(t.exportDefaultDeclaration(exportId));
+
+    callExpressionPath.scope.registerDeclaration(paths[0]);
+    callExpressionPath.scope.registerDeclaration(exportDefaultPath);
+
+    // this "creates" a new callExpressionPath, so it will be picked up again on the
+    // next visitor call. no need to continue.
+    return;
+  }
+
+  /**
+   * Normally, we assume the Component to be a identifier (e.g. block(Component)) that
+   * references a function. However, it's possible that the user passes an anonymous
+   * function (e.g. block(() => <div />)). In this case, we need to hoist the function
+   * declaration to the top of the scope and replace the anonymous function with the
+   * identifier of the hoisted function.
+   *
+   * ```js
+   * block(() => <div />)
+   *
+   * // becomes
+   *
+   * const anonymous = () => <div />;
+   *
+   * block(anonymous)
+   * ```
+   */
+
+  if (
+    t.isFunctionExpression(rawComponent) ||
+    t.isArrowFunctionExpression(rawComponent)
+  ) {
+    const name =
+      t.isFunctionExpression(rawComponent) && t.isIdentifier(rawComponent.id)
+        ? rawComponent.id.name
+        : publicComponent?.name || 'Anonymous';
+    const anonymousComponentId = getUniqueId(callExpressionPath, name);
+
+    /**
+     * Replaces function expression with identifier
+     * `block(() => ...)` to `block(anonymous)`
+     */
+
+    // TODO: check this
+    const paths = callExpressionPath.parentPath.parentPath?.insertBefore(
```

Comment:
I'm not sure if direct access to `parentPath` is safe here. I'd usually crawl towards the parent statement path and then insert it before. Needs testing.